### PR TITLE
Fixes #410 - Retrieve title fails in multiple cases

### LINKED ISSUE #410: Title do not show if space at the end of the URL
Hi,

If I paste the URL in the "add link" form with some extra spaces at the end of the url (when copied manually for example), the title doesn't auto fill. It seems to be a simple trim I guess.

Regards,
Tom

### PR DESCRIPTION
**Pull request by ArthurHoaro**
* `get_http_url()` renamed to `get_http_response()`.
  * Use the same HTTP context to retrieve response headers and content.
  * Follow HTTP 301 and 302 redirections to retrieve the title (default max 3 redirections).
  * Add `LinkUtils` to extract titles and charset.
  * Try to retrieve charset from HTTP headers first (new), then HTML content.
  * Use mb_string to re-encode title if necessary.

### CODE REVIEW

**Comment by ArthurHoaro** [comment]
> All checks have failed

[ ] Unit tests.

File: application/HttpUtils.php [hunk 1]
```diff
@@ -24,31 +24,66 @@
...
+ * Retrieve HTTP headers, following n redirections (temporary and permanent).
+ *
+ * @param string $url initial URL to reach.
+ * @param int $redirectionLimit max redirection follow..
+ *
+ * @return array
+ */
+function get_redirected_headers($url, $redirectionLimit = 3)
+{
+    $headers = @get_headers($url, 1);
```
Comment (ArthurHoaro) [inline application/HttpUtils.php]: I couldn't find a better way to pass the unit tests. PHPUnit seems to focus on the [failure/warning](https://travis-ci.org/shaarli/Shaarli/jobs/101131361) and ignores tests I made.
If anyone has a better idea about this, I'd be happy to take it.

**Comment by ArthurHoaro** [comment]
Updated! The initial PR contained regressions, avoided thanks to unit tests! 😊

File: application/HttpUtils.php (hunk 1 above)
Comment (virtualtam) [inline application/HttpUtils.php]: Stumbled upon this cool hub SO thread: [Reference - What does this symbol mean in PHP?](http://stackoverflow.com/questions/3737139/reference-what-does-this-symbol-mean-in-php); `@` is an [error control operator](http://php.net/manual/en/language.operators.errorcontrol.php) that allows hiding errors (which seems quite bad practice :p).

Looking at the PHP docs, it is possible to set [custom error handlers](http://php.net/manual/en/function.set-error-handler.php), that could be used to systematically convert PHP errors to exceptions, as suggested by [this article](http://anvilstudios.co.za/blog/2010/04/09/how-to-ignore-errors-in-a-custom-php-error-handler/) (via [this SO post](http://stackoverflow.com/a/4872354)).

Comment (ArthurHoaro) [inline application/HttpUtils.php]: That's interesting. I'll look further into it. However, the error handler is globally set and overrides the default error handler. 

>  It is important to remember that the standard PHP error handler is completely bypassed for the error types specified by error_types unless the callback function returns FALSE. [...]
> Also note that it is your responsibility to die() if necessary. If the error-handler function returns, script execution will continue with the next statement after the one that caused an error.

We need to be very careful about this.

File: tests/LinkUtilsTest.php [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+    {
+        $charset = 'x-MacCroatian';
+        $html = '<html><meta>stuff2</meta><meta charset="'. $charset .'"/></html>';
+        $this->assertEquals(strtolower($charset), html_extract_charset($html));
+    }
+
+    /**
+     * Test html_extract_charset() when the charset is not found.
+     */
+    public function testHtmlExtractInexistingCharset()
```
Comment (virtualtam) [inline tests/LinkUtilsTest.php]: Inexistent / NonExistent

File: tests/Url/UrlTest.php [hunk 3]
```diff
@@ -156,4 +156,22 @@ function testAddTrailingSlash()
         $this->assertEquals($strOn, add_trailing_slash($strOn));
         $this->assertEquals($strOn, add_trailing_slash($strOff));
     }
+
+    /**
+     * Test valid HTTP url.
+     */
+    function testIsUrlTrue()
```
Comment (virtualtam) [inline tests/Url/UrlTest.php]: `testUrlIsHttp()` / `testUrlIsNotHttp()` ?

File: tests/LinkUtilsTest.php (hunk 2 above)
Comment (ArthurHoaro) [inline tests/LinkUtilsTest.php]: Yep, that felt wrong when I wrote it.

File: application/HttpUtils.php (hunk 1 above)
Comment (virtualtam) [inline application/HttpUtils.php]: An additional (and a bit less hack-ish) possibility is to toggle PHP's error handling during tests, when calling the faulty methods, e.g.

```php
function testGetRedirectedHeaders()
{
    [...]
    list($headers, $url) = @get_redirected_headers('http://u.rl');
    [...]
}
```

Comment (ArthurHoaro) [inline application/HttpUtils.php]: Good idea! I'll go with it. I tried to do something with `set_error_handler()` but a lot of things can happen, and the code is not ready to catch exceptions everywhere.

**Comment by ArthurHoaro** [comment]
Thank you @virtualtam! Updated regarding your last comments.

**Comment by virtualtam** [comment]
You're welcome :)

**Pull request merged**